src/avt/Queries/Queries/avtXRayImageQuery.C

```diff
@@ -1081,26 +1219,26 @@ avtXRayImageQuery::Execute(avtDataTree_p tree)
             EXCEPTION1(VisItException, "There must be at least one bin.");
         }
 
+        int numBins = numLeaves / 2;
+
         //
         // Create the file base name.
         //
         std::stringstream baseName;
         const int cycle = GetInput()->GetInfo().GetAttributes().GetCycle();
-        bool keepTrying = true;
-        while (keepTrying)
+        if (filenameScheme == FAMILYFILES)
         {
-            keepTrying = false;
-            if (familyFiles)
+            auto keepTrying{true};
```

Comment:
remind me again...what does `auto` keyword do?

Reply:
This use of auto is really just a stylistic one. It doesn't solve any particular problem and there is not any fantastic reason to write `auto myvar{true}` instead of `bool myvar{true}`. The reason I use it is because I was told in the past by [Mr. C++ himself](https://cpp.chat/) that I should use `auto` unless I have a good reason not to.

What does `auto` do? It basically just copies the type over. So whatever the type of the object on the right is, that is the type that the compiler will deduce for the new object we are declaring on the left.

The following are some of my notes on `auto` and why it is great :)

---------------------------

note: auto makes my code more responsive to changes in the future
auto isn't just a lazy thing, it's a way of saying you want to preserve the type:
`auto x{myval}` will give x the same type as myval even if myval's type changes

**prefer auto to explicit types when declaring objects**
less verbose syntax :)
you can lose data if you aren't careful if you don't use auto

auto just copies the type over

note: you can lose data if you don't use auto - say a return type changes, auto will save you; otherwise the compiler will try to cast and you could lose data or have an even worse day

`auto myvar(static_cast<mytype>(foo))` is extremely clear and safe
better than `mytype myvar = foo;`. The second one could just mean we think `myvar` is going to be a `mytype`, while the first says I know that I want `myvar` to be `mytype`
also good for the future, if someone else reads the code there is nothing ambiguous about the first way.
only needed if we don't want what it returns and want something specific and different

Reply:
Hmmmm, I think I've messed up here though. It should be `auto myvar = myval`, because of initializer list behavior.

Reply:
In fact, I think I've done this wrong elsewhere. A quick search of our code shows 4 places where I've made this mistake. In each case it should instead be `auto myvar = myval` or `auto myvar(myval)`. I hate initializer lists >:(

Experts online suggest that I've made a mistake, but each of the cases should cause an error if that is the case. I think I'll change each of these just in case, even though everything works as intended in each case. Just to be safe.

Reply:
> The reason I use it is because I was told in the past by [Mr. C++ himself](https://cpp.chat/) that I should use `auto` unless I have a good reason not to.

IMHO, `auto myvar(another_var)` is a tad different then a literal value as in `auto myvar(true)`. The former does I guess provide some future proofing if type of `another_var` needs to change. And, you can't `auto`-ize `float x = 0`  (you need `auto x = 0.f`) nor even `float x = 3.3` (you need `auto x = 3.3f` otherwise x is `double`). On the other hand, having to go visit all the places where a variable is used after its type has been changed is, IMHO, not a bad thing. Strong typing is one reason C++ was favored over C and `auto` is a half-baked scheme (IMHO) to get around *some* of that.

And, I have no idea why prefering curly braces (`bool x{true}`) (e.g. initializer list mode) is better than parens `bool x(true)` (e.g. copy constructor) or, everywhere else in VisIt `bool x = true;`. We don't do `bool *x = new bool{true}` because that is a compiler error. Instead we do `bool *x = new bool(true);`. In fact, the curly braces don't even compile on `gcc` on macOS out of the box...

```
% cat foo.C
#include <new>
int main()
{
    bool x{true};
    int y(5);
    int *z = new int(7);
    auto q = 3.f;
    return 0;
}

% g++ foo.C -o foo
foo.C:4:11: error: expected ';' at end of declaration
    bool x{true};
          ^
          ;
```

If this style of local variable declaration is such that it is subconsciously programmed into your finger tips at this point (as `vi` commands are in mine), then I can understand how having to think to match existing code base in this regard is a little inconvenient. Otherwise, this feels to me a lot like stylistic departure from existing code base without good reason (I guess I am not easily convinced by C++ pundits) and I am not in favor of it.

I take it back, a long while ago I was taught `if (7 == x) ...` is way better than `if (x == 7) ...` and it has to do with chance of forgetting the second `=`. In the former, you would get a compiler error. In the later, it can be silently missed (though I think modern compilers do warn about that if they enounter it) which is actually sad because sometimes I want to assign *and* test in the same statement as in `if (x = y) ...` and warning then is problematic. I really wanted to do that.

Reply:
I've actually removed this instance of `auto` in an ensuing commit.

You are also very right in saying
> `auto myvar(another_var)` is a tad different then a literal value as in `auto myvar(true)`

The former is useful sometimes as discussed. The latter (doing type deduction on primitives) isn't all that useful, but it also is unlikely to hurt.

On the subject of using curly braces versus parens for initialization... here are my (unasked for) notes on the subject (yes I did write my notes in markdown originally, please don't judge me!):

-----------------------------------

### Distinguish () and {} when creating objects
{} is the standard way to initialize stuff
but don't do it with auto! you'll be in trouble in many versions of cpp
narrowing conversions (double->int truncation) are errors with brace initialization in modern cpp

parens can be scary, if you call w/ args then probably good but `Widget w()` will make a function called `w` that returns a Widget. This works as expected if you do with {}

*you can get different things if you call with {} or (), since {} are initializer lists*

recommend to use {} to initialize *unless* there is a constructor for your object that takes an init. list, in which case you should use () instead.

braces use std::initializer_list overload if available, parens never do

empty braces call default constructor **NOT** the empty init list
want to call w/ an empty init list, do nested braces {{}}

----------------------------

This post discusses why one should use braces for initialization over `=`: https://stackoverflow.com/questions/18222926/what-are-the-advantages-of-list-initialization-using-curly-braces

Returning from the tangent:

> If this style of local variable declaration is such that it is subconsciously programmed into your finger tips at this point (as vi commands are in mine), then I can understand how having to think to match existing code base in this regard is a little inconvenient. Otherwise, this feels to me a lot like stylistic departure from existing code base without good reason (I guess I am not easily convinced by C++ pundits) and I am not in favor of it.

Using auto here was not necessary (and wrong!), hence why I removed it. I've been trying to only use when it solves a specific problem, and I'll try harder on that front.

Reply:
> https://stackoverflow.com/questions/18222926/what-are-the-advantages-of-list-initialization-using-curly-braces

Oh, thanks for that. That is a good ref and I have encountered such situations in the past but never had the full picture. That makes things make a bit more sense.

Reply:
It should be noted that his final example is not quite correct:
`auto z1 {99};   // z1 is an int` is not _always_ true. It is dependent on the compiler and cpp standard you are using. It could be an initializer list instead if you get unlucky. That's why I went ahead and changed all instances of this in our code today.

Reply:
> auto z1 {99}; // z1 is an int is not always true

At the risk of splitting hairs, isn't C++ spec going to *always* treat `99` (with no suffix) as `int`? I mean `99U` would be `unsigned int` (https://en.cppreference.com/w/cpp/language/integer_literal).